frontends/p4/alias.h

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2019 VMware, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _FRONTENDS_P4_ALIAS_H_
+#define _FRONTENDS_P4_ALIAS_H_
+
+/*
+ * A simple conservative syntactic alias analysis.  Two things may
+ * alias if they refer to objects with the same name.
+ * This pass may be run early in the front end, before enough information is present (eg.
+ * def-use information) to do a precise alias analysis.
+ */
+
+#include "ir/ir.h"
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/p4/methodInstance.h"
+
+namespace P4 {
+
+/// This class represents the path to a location.
+/// Given a struct S { bit a; bit b; } and a variable S x;
+/// a path can be x.a, or just x.  An array index is represented as a
+/// number (encoded as a string) or as "*", denoting an unknown index.
```

Comment:
What about a slice?

Reply:
I don't see slices handled, I should probably handle these as well. Currently this will approximate a slice with the whole value that is sliced.

Reply:
Actually slice is handled by Operation_Ternary.